zkvm/src/prover.rs

```diff
@@ -44,8 +57,68 @@ impl<'a, 'b> Delegate<r1cs::Prover<'a, 'b>> for Prover<'a, 'b> {
     }
 }
 
-impl RunTrait for RunProver {
+impl<'a, 'b> Prover<'a, 'b> {
+    pub fn build_tx<'g>(
+        program: Vec<Instruction>,
+        version: u64,
+        mintime: u64,
+        maxtime: u64,
+        bp_gens: &'g BulletproofGens,
+    ) -> Result<Tx, VMError> {
+        let mut r1cs_transcript = Transcript::new(b"ZkVM.r1cs");
+        let pc_gens = PedersenGens::default();
+        let cs = r1cs::Prover::new(bp_gens, &pc_gens, &mut r1cs_transcript);
+
+        let mut prover = Prover {
+            signtx_keys: Vec::new(),
+            cs: cs,
+        };
+
+        let vm = VM::new(
+            version,
+            mintime,
+            maxtime,
+            ProverRun::new(&program),
+            &mut prover,
+        );
+
+        let (txid, _) = vm.run()?;
+
+        // Sign txid
+        let mut signtx_transcript = Transcript::new(b"ZkVM.signtx");
+        signtx_transcript.commit_bytes(b"txid", &txid.0);
+        let signature = Signature::sign_aggregated(&mut signtx_transcript, &prover.signtx_keys);
+
+        // Generate the R1CS proof
+        let proof = prover.cs.prove().map_err(|_| VMError::InvalidR1CSProof)?;
+
+        // Encode program into bytecode
+
+        // TBD: determine program capacity
+        let mut bytecode = Vec::new();
+        program.iter().for_each(|p| p.encode(&mut bytecode));
+
+        Ok(Tx {
+            version,
+            mintime,
+            maxtime,
+            signature,
+            proof,
+            program: bytecode,
+        })
+    }
+}
+
+impl ProverRun {
+    fn new(program: &Vec<Instruction>) -> Self {
+        ProverRun {
+            program: program.clone().into(),
```

Comment:
Can we design it, so that instead of having two full copies of programs: one for consumption by VM, another for converting into bytecode, we consume the stream of `Instruction` and simultaneously encode them into a bytecode buffer?

We can do so with a hook like `delegate.will_execute_instruction(&Instruction)`.

Reply:
sounds good to me! would that fit best inside the call to `vm::´step` before we enter the big match block?

Reply:
or I imagine we could serialize the instruction when we call `next_instruction` on the prover?